02_bubble/ailyntang/sort_test.go

```diff
@@ -0,0 +1,37 @@
+package main
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSortEdgeCases(t *testing.T) {
+	assert := assert.New(t)
+
+	assert.Equal([]int{}, bubble([]int{}), "empty array as input")
+	assert.Equal([]int{4}, bubble([]int{4}), "one integer as input")
+	assert.Equal([]int{0}, bubble([]int{0}), "0 as input")
```

Comment:
I would add one more test to the edge cases:
A slice with 2 elements

This is one more boundary condition in your code. In the first one (empty slice), the outer loop should not be entered. In the second, the inner loop should not be entered. For a 2-element slice, the inner loop should be entered. I'd use an input that is not sorted to ensure you do the sorting.

I would also suggest removing some of the other tests - all negative and all positive do not work the code under test any differently. A mixture of positive and negative covers both of those other tests, and also validates that you are using the correct types and not mishandling negative numbers (although in a statically-typed language like Go, the compiler will typically pick up those sorts of problems).

Reply:
Thanks for the explanation about testing the different `for` loops, and about the redundancy of the other tests. 

Both make a lot of sense. I've updated the code for this.